### PR TITLE
API: deprecate undocumented functions

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @rgommers,

Here I share a PR with a next batch of deprecated functions from https://github.com/numpy/numpy/issues/12385. In this PR deprecated functions are:
- `recfromcsv`
- `recfromtxt`
- `disp`
- `get_array_wrap`
- `maximum_sctype`
- `deprecate`
- `deprecate_with_doc`

(they aren't used except for dedicated tests so dropping them for good should be just removing them)

**[EDIT]**
Also this PR un-deprecate `byte_bounds` as agreed due to downstream impact.

### CODE REVIEW

**Comment by seberg** [comment]
Can we just avoid `@deprecate` (we can deprecate it also, or just remove it TBH, I don't see why downstream needs our helper, they can just vendor it).
It doesn't provide a nice message or docs update.  I am not even sure it provides the right stacklevel (which leads to issues like gh-24015, although there the stacklevel is just wrong stacklevel).

**Comment by mtsokol** [comment]
> Can we just avoid `@deprecate` (we can deprecate it also, or just remove it TBH, I don't see why downstream needs our helper, they can just vendor it). It doesn't provide a nice message or docs update. I am not even sure it provides the right stacklevel (which leads to issues like [gh-24015](https://github.com/numpy/numpy/issues/24015), although there the stacklevel is just wrong stacklevel).

Hi @seberg, 
Sure! I can remove it, just to explain - I used `@deprecate` in previous PR #23830.
Then, as I understand, deprecation should happen with a simple
```python
warnings.warn("...", DeprecationWarning)
```
at the top of deprecated function, is it correct? Or is there another approach?

**Comment by seberg** [comment]
For these functions yes (they don't have a `array_function_dispatch` decoration).  You also need a `stacklevel=2` if the function is directly called by the user (as is the case).  We normally put a `# Deprecated 2023-07-10, NumPy 2.0` above it and a `(deprecated in NumPy 2.0)` in the warning itself so that users have an idea for how new the warning is it at least.

**Comment by ngoldbaum** [comment]
It would be good to suggest migration paths for users of these functions. If no good migration path is available, we may want to reconsider deprecating. I would search for downstream usages to get a feeling for how functions you'd like to deprecate are used in the real world. For example, the [nilearn library seems to have some usages of `recfromcsv`](https://github.com/search?q=repo%3Anilearn%2Fnilearn%20recfromcsv&type=code) in their tests and documentation.

**Comment by mtsokol** [comment]
> It would be good to suggest migration paths for users of these functions in the deprecation. If no good migration path is available, we may want to reconsider deprecating. I would search for downstream usages to get a feeling for how functions you'd like to deprecate are used in the real world. For example, the [nilearn library seems to have some usages of `recfromcsv`](https://github.com/search?q=repo%3Anilearn%2Fnilearn%20recfromcsv&type=code) in their tests and documentation.

Hi @ngoldbaum,
Sure! I see that previous deprecation PR had a downstream impact https://github.com/numpy/numpy/pull/23830#issuecomment-1629208691 so I think it will be helpful to also provide substitutes for deprecated functions in the deprecation message.
Let me maybe prepare a summary of all deprecates/plans from #12385 and NEP 52 in a single table with proposed replacements and downstream impacts and then we can move forward with it. Does it make sense?

**Comment by rgommers** [comment]
> Let me maybe prepare a summary of all deprecates/plans from #12385 and NEP 52 in a single table/memo with proposed replacements and downstream impacts and then we can move forward with it. Does it make sense?

I like that plan - much nicer than having separate release note snippets for each deprecation. Perhaps it will end up looking like this page: https://numpy.org/devdocs/reference/array_api.html#table-of-differences-between-numpy-array-api-and-numpy?

If there's a simple one-for-one replacement, it's good to have that in the deprecation message too.

**Comment by mtsokol** [comment]
> For these functions yes (they don't have a `array_function_dispatch` decoration). You also need a `stacklevel=2` if the function is directly called by the user (as is the case). We normally put a `# Deprecated 2023-07-10, NumPy 2.0` above it and a `(deprecated in NumPy 2.0)` in the warning itself so that users have an idea for how new the warning is it at least.

Hi @seberg! I deprecated `deprecate` and `deprecate_with_doc` and moved to manual warning with `DeprecationWarning`.

**Comment by rgommers** [comment]
The test failures are real, all due to the same leftover usage of `np.deprecate` in a test it looks like.

**Review by ngoldbaum — COMMENTED** [review]
While looking at the current devdocs I noticed that the `deprecate` decorator is adding a note to the docstrings of the deprecated functions that they are deprecated. Maybe it would be a good idea to manually add a note to the docstrings of deprecated functions that they were deprecated in NumPy 2.0?

File: numpy/core/numerictypes.py [hunk 1]
```diff
@@ -168,6 +168,15 @@ def maximum_sctype(t):
     <class 'numpy.float128'> # may vary
 
     """
+
+    # Deprecated in NumPy 2.0, 2023-07-11
+    warnings.warn(
+        "`maximum_sctype` is deprecated. "
```
[the remark this answers is not in the document]
Comment (rgommers) [reply numpy/core/numerictypes.py:174]: It doesn't seem all that useful to me - the linked issue points at https://stackoverflow.com/questions/48410604/find-maximum-integer-type-at-runtime-in-numpy/48412971#48412971, but the original question there doesn't really make much sense, I can't think of any reason why you'd want your code to dynamically determine the largest unsigned integer like that (it's always `uint64`) nor why you would want the answer to change in the unlikely case NumPy adds or remove a dtype.

I don't think any of the `sctype` functions have enough value to have them stay, as touched upon in gh-17325 and explicitly mentioned in https://numpy.org/neps/nep-0052-python-api-cleanup.html#cleaning-up-the-main-namespace (no one disagreed with that I believe).

[the remark this answers is not in the document]
Comment (ngoldbaum) [reply numpy/core/numerictypes.py:174]: Sure, not advocating to keep it, just trying to think about how you'd replace it if you were using it. 

This one does seem to have [downstream]( https://github.com/search?q=np.maximum_sctype%28np.float%29&type=code) [usages](https://github.com/astropy/astropy/blob/bb2b6225278bef53c92ff8ca10fa1ddefa3ab2cf/astropy/nddata/bitmask.py#L22), so I think it's worth thinking about what to suggest as a replacement.

[the remark this answers is not in the document]
Comment (rgommers) [reply numpy/core/numerictypes.py:174]: That works, thanks. At least it should cover all the example usages I've looked at.

File: numpy/core/arrayprint.py [hunk 2]
```diff
@@ -1716,6 +1716,15 @@ def set_string_function(f, repr=True):
     'array([0, 1, 2, 3])'
 
     """
+
+    # Deprecated in NumPy 2.0, 2023-07-11
+    warnings.warn(
+        "`set_string_function` is deprecated. "
+        "(deprecated in NumPy 2.0)",
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply numpy/core/arrayprint.py:1746]: Sure! Added

File: numpy/core/numerictypes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtsokol) [reply numpy/core/numerictypes.py]: Sure! I included recommended message.

File: numpy/lib/utils.py [hunk 3]
```diff
@@ -1076,6 +1102,15 @@ def safe_eval(source):
     ValueError: malformed node or string: <_ast.Call object at 0x...>
 
     """
+
+    # Deprecated in NumPy 2.0, 2023-07-11
+    warnings.warn(
+        "`safe_eval` is deprecated. "
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply numpy/lib/utils.py]: Sure, I added this recommendation.

**Comment by mtsokol** [comment]
> While looking at the current devdocs I noticed that the `deprecate` decorator is adding a note to the docstrings of the deprecated functions that they are deprecated. Maybe it would be a good idea to manually add a note to the docstrings of deprecated functions that they were deprecated in NumPy 2.0?

Hi @ngoldbaum,
`deprecate` adds `<func_name> is deprecated! Use <replacement> instead.`, I can add this preamble to every function's stringdoc that I deprecated - should I add it?

**Comment by seberg** [comment]
The nicest thing to to should be adding:
```
.. deprecated:: 2.0
   Additional guidance on replacement if you have it.  (can be left empty)
```
I guess after the initial sentence probably, although you can see how it is used otherwise (not sure how often it is used for sentences).

That will render as a nice box.

**Comment by mtsokol** [comment]
Hi! This PR is ready for another round of review (Four CI checks failed due to `Do you want to continue? [Y/n] Abort.` statement on installation).

**Review by ngoldbaum — APPROVED** [review]
This looks great and should be a good template for the work necessary for future python API deprecations.

**Comment by ngoldbaum** [comment]
I opened https://github.com/numpy/numpy/pull/24188 which should hopefully avoid the test failures you're seeing on Cirrus.

**Review by rgommers — APPROVED** [review]
LGTM too!

**Pull request merged**